internal/router/router.go

```diff
@@ -228,29 +276,23 @@ func (r *Router) serveIds(w http.ResponseWriter, req *http.Request) error {
 	if key == "" {
 		return errgo.WithCausef(nil, params.ErrNotFound, "")
 	}
-	fullySpecified := url.Series != "" && url.Revision != -1
 	handler := r.handlers.Id[key]
-	if handler == nil || idHandlerNeedsResolveURL(req) {
-		// If it's not an id handler, it's a meta endpoint, so
-		// we always want a resolved URL. Otherwise we leave the
-		// URL unresolved for cases where the id may validly not
-		// exist (for example when uploading a new charm).
-		if err := r.resolveURL(url); err != nil {
-			// Note: preserve error cause from resolveURL.
-			return errgo.Mask(err, errgo.Any)
-		}
-	}
 	if handler != nil {
 		req.URL.Path = path
-		err := handler(url, fullySpecified, w, req)
+		err := handler(url, w, req)
 		// Note: preserve error cause from handlers.
 		return errgo.Mask(err, errgo.Any)
 	}
+	rurl, err := r.resolveURL(url)
```

Comment:
Also, can we move this call after the `key != "meta/" && key != "meta"` check?
This way, we avoid hitting the db if the request path is not valid.